enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+In this case we are making an exception since we know ahead of time that the structure is not needed to be changed, the value is an enum,
+and that the enum values will be static and not be changed in the future.
+The annotation approach in this case means we can write generic code to solve the authority issue across many objects,
+without adding a concrete API to every Machine API custom resource definition.
+
+#### Authority of resources and controller reconciliations
+
+To ensure that only one controller acts on any resource (Machine, MachineSet etc) at any one time,
+a new annotation `machines.openshift.io/authoritative-api` will be implemented/
+The new annotation will be applied to a Machine API version of a resource to mark it as authoritative.
+
+The accepted values for the annotation will be either `machine.openshift.io` or `cluster.x-k8s.io` (the respective API groups for Machine API and Cluster API).
+If the value is not present, it will be added by the Machine API resource controller and defaulted to `machine.openshift.io`.
+
+When the Machine API controller attempts to reconcile an object, it must first check for the presence of the annotation.
+If it is not present, the annotation will be added and the controller will requeue the object for later processing.
+
+We will implement predicates/event filters for both Machine API and Cluster API controllers.
+
+Machine API controllers will be allowed to reconcile any object that is either:
+
+* Missing the authoritative annotation (as today)
+* Has the authoritative annotation set with the `machine.openshift.io` value
+
+Cluster API controllers will be allowed to reconcile any object that either:
+
+* Has a Machine API equivalent AND has the authoritative annotation set with the `cluster.x-k8s.io` value
+* Does not have a Machine API equivalent
+
+By following the above rules, we should be able to ensure that the authoritative annotation is present in only one location and that the reconciliation of the objects is well defined based on the presence of this annotation.
+
+### Implementation Details/Notes/Constraints
+
+#### Synchronisation of resources
+
+To ensure that resources are synchronised, we will implement a new controller `machine-api-synchronisation-controller`.
+This controller will be responsible for translation of Machine API resources to Cluster API resources, and vice-versa.
+
+This controller will be responsible for creating Cluster API versions of Machine API resources, when they do not exist.
+It will observe the authoritative annotation described [above][authoritative-annotation] to synchronise between the two resources,
+overwriting anything in the non-authoritative resource that does not match with the authoritative resource.
+
+When errors occur during synchronisation, the controller will write conditions to the Machine API resource to report the failure.
```

Comment:
Correct to a degree. Any new fields we add will be on the MAPI side. However, we can add a Condition to the status of the CAPI side, in theory, without it affecting the operation of the CAPI resource itself. So if there is an issue that we want to report, we can do so via conditions

Reply:
Conditions... but not other status fields? (Like the generation of the MAPI mirror when it is authoritative, for symmetry with what we're planning to do to the MAPI object's status?)

Reply:
Yes, my intention for CAPI is just to use the CAPI conditions addition, and all other status, for sync in both directions, will be on the MAPI resource. Long term we don't want to carry patches on the CAPI controllers and doing anything more than this would mean carries